Relative dates in deadlines and scheluded for before and after

dancn opened the pull request:
Backward compatible improvement, org-read-date support a super-set of
the previous function org-time-string-to-absolute.

Should be also an easy solution for #59.

alphapapa commented:
Thanks, this looks like a nice improvement.  I will look at it more closely when I have some time, hopefully soon.

nivekuil commented:
+1, I use this patch locally and it's quite convenient.

pkazmier commented:
Yes, I, too, think this would be incredibly useful, so much so, I submitted the exact same PR without realizing this one existed!

dancn commented:
After one year I re-based the branch for my convenience, hopefully will also speed up the review and merge.

real-or-random commented:
It would be great to see this merged!

alphapapa commented:
Thanks for the ping and your patience.

alphapapa commented on README.org:
```diff
@@ -191,7 +191,7 @@ These selectors take one argument alone, or multiple arguments in a list.
 +  =:category= :: Group items that match any of the given categories.  Argument may be a string or list of strings.
 +  =:children= :: Select any item that has child entries.  Argument may be ~t~ to match if it has any children, ~nil~ to match if it has no children, ~todo~ to match if it has children with any to-do keywords, or a string to match if it has children with certain to-do keywords.  You might use this to select items that are project top-level headings.  Be aware that this may be very slow in non-dai ...
 +  =:date= :: Group items that have a date associated. Argument can be =t= to match items with any date, =nil= to match items without a date, or =today= to match items with today’s date. The =ts-date= text-property is matched against.
```
Should other selectors, like `:date`, also support this?

dancn replied:
> Thanks for your patience. I think this shouldn't need much more before it can be merged.

Thanks for the review!
 
> Also, would you add tests for this, please? (i.e. in the test suite, not just that Org file)

Just added a quick test (not pushed yet). It fails, I guess that org-read-date is not aware of org-super-agenda-test-date so relative date specification can not be tested without further mocks. Do you have some suggestion on how to do that?

:date does not support arbitrary dates yet (only static: any, none, today), so relative ones are a double step!

alphapapa replied:
> > Thanks for your patience. I think this shouldn't need much more before it can be merged.
> 
> Thanks for the review!

FYI, feel free to ping me more often in the future.  I didn't intend to ignore this for a year; I just hadn't thought about working on org-super-agenda for a while, since I didn't need to change anything in it for my own sake, and I have a number of other projects that occupy my free time.

> > Also, would you add tests for this, please? (i.e. in the test suite, not just that Org file)
> 
> Just added a quick test (not pushed yet). It fails, I guess that org-read-date is not aware of org-super-agenda-test-date so relative date specification can not be tested without further mocks. Do you have some suggestion on how to do that?

I guess you'll have to look in the Org source code, find the appropriate function, and redefine it locally in the test using `cl-letf`.

> :date does not support arbitrary dates yet (only static: any, none, today), so relative ones are a double step!

As you can tell, I haven't worked on this package for a while.  :)  We should probably fix that too, though maybe not in this PR.

alphapapa commented on org-super-agenda.el:
```diff
@@ -437,8 +437,9 @@ match items that have no deadline), `past` (to match items with a
-`org-time-string-to-absolute' can process."
+DATE', where DATE is a date string that `org-read-date' can
+process.  Note that relative dates are supported, e.g. `before
++3d' means in the next two days."
```
The `+3d` here should be in a string, right?

dancn replied:
Yes, do you think that is not clear?

alphapapa replied:
The part in the backtick-apostrophe quote indicates Lisp code, i.e. like `(:scheduled (before "+3d"))`.  Without the string, it wouldn't be valid, e.g. `(:scheduled (before +3d))`.  So the `+3d` should appear as a string, I think.

alphapapa commented on org-super-agenda.el:
```diff
@@ -474,8 +475,9 @@ Argument can be `t' (to match items scheduled for any date),
-`org-time-string-to-absolute' can process."
+DATE', where DATE is a date string that `org-read-date' can
+process.  Note that relative dates are supported, e.g. `before
++3d' means in the next two days."
```
See previous comment.  :)

alphapapa reviewed (changes requested):
Thanks for your patience.  I think this shouldn't need much more before it can be merged.

Also, would you add tests for this, please?  (i.e. in the test suite, not just that Org file)

Commit:
Relative dates in deadlines and scheluded for before and after

Backward compatible improvement, org-read-date support a super-set of
the previous function org-time-string-to-absolute.

The tests are working, but hugly hacks to force org-read-date to compute
a relative date to a fixed date.

dancn commented:
Another try here! I found out out to get relative date to a fixed date... the test code is not very lispy but works and covers the obvious cases.

eugr commented:
Any updates on this? Anything preventing this from being merged? This would be a mega-useful feature as there is no good workaround for this currently.

DominikMendel commented:
I did this as a local work around to adjust for the number of days. Hope it helps anyone else. It's a little janky. 
```
(defun my/get-time-string-today-offset (dayOffSet)
  (format-time-string "%Y-%m-%d %H:%M" (+ (string-to-number (format-time-string "%s" (current-time))) (* 86400 dayOffSet))))
```

And call it like this:
```
  (let ((date (my/get-time-string-today-offset 5)))
....
`( ...
:scheduled (after ,date)
)
```

alphapapa commented:
> Any updates on this? Anything preventing this from being merged? This would be a mega-useful feature as there is no good workaround for this currently.

@eugr What prevents it from being merged is chiefly my lack of time to work on it.  I don't have as much to spend on this project as I used to.  

Besides that, there are issues of code quality, tests to be reviewed, etc. that must be addressed before it can be merged.  Some of them I have mentioned to the PR's author in comments, so in that respect, the PR is waiting on him to address them, and then on me to re-review and confirm that it's appropriate for merging. 

#225 is also a possibility for merging, but it also needs to be polished and re-reviewed, and its author has not responded to my last comment.

Of course, if either of these PRs works for you, you're free to use their code as-is right now.
